builder/hyperv/common/step_set_first_boot_device.go

```diff
@@ -0,0 +1,140 @@
+package common
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/hashicorp/packer/helper/multistep"
+	"github.com/hashicorp/packer/packer"
+)
+
+type StepSetFirstBootDevice struct {
+	Generation      uint
+	FirstBootDevice string
+}
+
+func ParseBootDeviceIdentifier(deviceIdentifier string, generation uint) (string, uint, uint, error) {
+
+	captureExpression := "^(FLOPPY|IDE|NET)|(CD|DVD)$"
+	if generation > 1 {
+		captureExpression = "^((IDE|SCSI):(\\d+):(\\d+))|(DVD|CD)|(NET)$"
+	}
+
+	r, err := regexp.Compile(captureExpression)
+	if err != nil {
+		return "", 0, 0, err
+	}
+
+	// match against the appropriate set of values.. we force to uppercase to ensure that
+	// all devices are always in the same case
+
+	identifierMatches := r.FindStringSubmatch(strings.ToUpper(deviceIdentifier))
+	if identifierMatches == nil {
+		return "", 0, 0, fmt.Errorf("The value %q is not a properly formatted device or device group identifier.", deviceIdentifier)
+	}
+
+	switch {
+
+	// CD or DVD are always returned as "CD"
+	case ((generation == 1) && (identifierMatches[2] != "")) || ((generation > 1) && (identifierMatches[5] != "")):
+		return "CD", 0, 0, nil
+
+	// generation 1 only has FLOPPY, IDE or NET remaining..
+	case (generation == 1):
+		return identifierMatches[0], 0, 0, nil
+
+	// generation 2, check for IDE or SCSI and parse location and number
+	case (identifierMatches[2] != ""):
+		{
+
+			var controllerLocation int64
+			var controllerNumber int64
+
+			// NOTE: controllerNumber and controllerLocation cannot be negative, the regex expression
+			// would not have matched if either number was signed
+
+			controllerNumber, err = strconv.ParseInt(identifierMatches[3], 10, 8)
+			if err == nil {
+
+				controllerLocation, err = strconv.ParseInt(identifierMatches[4], 10, 8)
+				if err == nil {
+
+					return identifierMatches[2], uint(controllerNumber), uint(controllerLocation), nil
+
+				}
+
+			}
+
+			return "", 0, 0, err
+
+		}
+
+	// only "NET" left on generation 2
+	default:
+		return "NET", 0, 0, nil
+
+	}
+
+}
+
+func (s *StepSetFirstBootDevice) Run(ctx context.Context, state multistep.StateBag) multistep.StepAction {
+
+	driver := state.Get("driver").(Driver)
+	ui := state.Get("ui").(packer.Ui)
+	vmName := state.Get("vmName").(string)
+
+	if s.FirstBootDevice != "" {
```

Comment:
all of this looks good, but again I'd like to see some tests for the logic. Here's an example of a test file. If you've not used golang testing before, you just need to add _test to the end of the name and plop it into the same directory (so here you'd be creating a new file named builder/hyperv/common/step_set_first_boot_device_test.go) 

Here's an example of a step test. 

https://github.com/hashicorp/packer/blob/master/builder/hyperv/common/step_create_vm_test.go

That example also shows how to put the mock driver into state so that it can be used to exercise the full step.

Reply:
Testing functions added.